keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,448 @@
+The series reported will be consistent with `kube_node_status_allocatable` to simplify queries that compare used capacity and available capacity. After a pod has reached the Completed lifecycle (as described above) state, the series will not be reported. If the value of the resource is zero, that resource will not report a series. They all share the form `kube_(running_|)pod_resource_(requests|limits)`.
+
+Note on naming: The Prometheus convention for a metric is to allow the sum of a metric to make sense. However, in the case of Kubernetes resources we are attempting to make discoverability of the reported metrics (which are unbounded in number as they are extensible) a key goal for admins. Most admins will understand CPU and memory usage, but would have to resort to metrics name wildcard regexes to find the other values if we give each resource its own metrics name. Also, because Kube extension resources have characters disallowed in metrics names, admins would continually have to transform `myresource.io/foo-bar` into `myresource_io_foo_bar` in order to locate names which places an undue burden on the user. So for the proposed metrics and `kube_node_status_allocatable` we instead place all types resources under the same metrics name with a unit and resource label.
+
+* `kube_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time.
+  * The value is the quantity of the request in `spec.containers.resources.requests`.
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `unit` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource or if it is an extension resource that does not have a detectable unit.
+* `kube_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time.
+  * The value is the quantity of the request in `spec.containers.resources.limits`.
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `node` - the node the resource is scheduled to, or empty if not yet scheduled
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `unit` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+
+The following metrics will be reported for pods that are in the `Running` phase and follow the same rules as above, but are not reported when the pod is `Pending` and can be used to calculate live resources. They all share the form `kube_running_pod_resource_(requests|limits)[_<resource>_<unit>]`.
```

Comment:
Should this be reported by the kubelet instead of by the scheduler?  It depends whose perspective you want on running resources...  For example, 

Lets say we have a kubelet that is a release behind the scheduler.  The scheduler takes pod overhead into account when computing pod requests, but the kubelet does not.  Do we want to measure what the kubelet has _applied_ to the running pod (without pod overhead), or what the scheduler has _accounted_ for the running pod (with pod overhead)?  Does consistency between the two matter?

I'm not sure the above case can actually happen, as the node has to advertise a runtimeclass with pod overhead... But I think the general question stands.

Reply:
I think we have to pick one of two places to report them no matter what.  The scheduler will make a placement decision regardless of the kubelet's version today, and you could argue that if placement is completely incorrect based on a gap between kubelet and scheduler, it is a bug resolved by making the scheduler aware of the gap.  Another practical consideration is that most admins will be making global decisions about usage, as will the scheduler, so having the node report running metrics has a period of time where the metric is incorrect (specifically - from scheduler's perspective the metric set by kubelet would be incorrect from the time spec.nodeName is set to the time the kubelet acknowledges the pod, whereas if the scheduler sets it both are consistent at sync time).  I would argue scheduler consistency is more useful to admins.

If we have implications that are per kubelet version, having a separate metric which tracks what the kubelet believes its resource overhead / total is might be more practical.